### PR TITLE
fix(iot-dev): Fix bugs with unnsuccessful synchronous operations

### PR DESCRIPTION
**Pull request by timtay-microsoft**
- Fix bug where synchronous operation did not throw an exception if it received an unsuccessful status code
- Fix bug where AMQP connections did not map AmqpLinkMessageSizeExceededException to REQUEST_ENTITY_TOO_LARGE
- Fix stack traces for synchronous methods not including the synchronous method call itself

Added some e2e tests to cover these scenarios. One test to check how the SDK responds to sending a message that exceeds IoT Hub's message size limit and one test to check how the SDK responds to sending a reported properties update with an out of date version.

### CODE REVIEW

**Comment by timtay-microsoft** [comment]
Example stacktrace before this fix
```java
com.microsoft.azure.sdk.iot.device.exceptions.IotHubClientException: Received an unsuccessful operation error code from the service: PRECONDITION_FAILED
	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.onMessageReceived(IotHubTransport.java:324)
	at com.microsoft.azure.sdk.iot.device.transport.mqtt.MqttIotHubConnection.onMessageArrived(MqttIotHubConnection.java:488)
	at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.messageArrived(Mqtt.java:364)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.deliverMessage(CommsCallback.java:514)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.handleMessage(CommsCallback.java:417)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.run(CommsCallback.java:214)
```


Example stacktrace after this fix
```java
com.microsoft.azure.sdk.iot.device.exceptions.IotHubClientException: Received an unsuccessful operation error code from the service: PRECONDITION_FAILED
	at com.microsoft.azure.sdk.iot.device.InternalClient.updateReportedProperties(InternalClient.java:450)
	at com.microsoft.azure.sdk.iot.device.InternalClient.updateReportedProperties(InternalClient.java:396)
	at tests.integration.com.microsoft.azure.sdk.iot.iothub.twin.TwinTests.sendMultipleReportedPropertiesChecksForVersion(TwinTests.java:458)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at tests.integration.com.microsoft.azure.sdk.iot.helpers.rules.BasicTierHubOnlyTestRule$IgnorableStatement.evaluate(BasicTierHubOnlyTestRule.java:36)
	at tests.integration.com.microsoft.azure.sdk.iot.helpers.rules.ContinuousIntegrationTestRule$IgnorableStatement.evaluate(ContinuousIntegrationTestRule.java:36)
	at tests.integration.com.microsoft.azure.sdk.iot.helpers.rules.DeviceProvisioningServiceTestRule$IgnorableStatement.evaluate(DeviceProvisioningServiceTestRule.java:36)
	at tests.integration.com.microsoft.azure.sdk.iot.helpers.rules.DigitalTwinTestRule$IgnorableStatement.evaluate(DigitalTwinTestRule.java:37)
	at tests.integration.com.microsoft.azure.sdk.iot.helpers.rules.ErrInjTestRule$IgnorableStatement.evaluate(ErrInjTestRule.java:37)
	at tests.integration.com.microsoft.azure.sdk.iot.helpers.rules.FlakeyTestRule$IgnorableStatement.evaluate(FlakeyTestRule.java:36)
	at tests.integration.com.microsoft.azure.sdk.iot.helpers.rules.IotHubTestRule$IgnorableStatement.evaluate(IotHubTestRule.java:36)
	at tests.integration.com.microsoft.azure.sdk.iot.helpers.rules.MultiplexingClientTestRule$MultiplexingClientTestStatement.evaluate(MultiplexingClientTestRule.java:29)
	at tests.integration.com.microsoft.azure.sdk.iot.helpers.rules.StandardTierHubOnlyTestRule$IgnorableStatement.evaluate(StandardTierHubOnlyTestRule.java:36)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:288)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:282)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:750)
Caused by: com.microsoft.azure.sdk.iot.device.exceptions.IotHubClientException: Received an unsuccessful operation error code from the service: PRECONDITION_FAILED
	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.onMessageReceived(IotHubTransport.java:324)
	at com.microsoft.azure.sdk.iot.device.transport.mqtt.MqttIotHubConnection.onMessageArrived(MqttIotHubConnection.java:488)
	at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.messageArrived(Mqtt.java:364)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.deliverMessage(CommsCallback.java:514)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.handleMessage(CommsCallback.java:417)
	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.run(CommsCallback.java:214)
	... 1 more
```

Note that in the before case, all the methods in the stacktrace are implementation details of the SDK and don't tell the user what method they called that caused this error. With the fix, the topmost method in the stacktrace is the synchronous method that the user called that caused this exception.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/exceptions/IotHubClientException.java [hunk 1]
```diff
@@ -47,7 +47,7 @@ public IotHubClientException(IotHubStatusCode statusCode, Exception nestedExcept
 
     public IotHubClientException(IotHubStatusCode statusCode, String errorMessage, Exception nestedException)
     {
-        super(nestedException);
+        super(errorMessage, nestedException);
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/exceptions/IotHubClientException.java:50]: Forgot to pass the error message to the super class alongside the nested exception

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java [hunk 2]
```diff
@@ -830,7 +839,7 @@ public void invokeCallbacks()
             IotHubClientException clientException = null;
             if (status != OK)
             {
-                clientException = new IotHubClientException(status);
+                clientException = new IotHubClientException(status, "Received an unsuccessful operation error code from the service: " + status);
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:842]: Adding an error message here for user context

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java [hunk 3]
```diff
@@ -1450,15 +1459,7 @@ private void handleMessageException(IotHubTransportPacket packet, TransportExcep
             log.warn("The device operation timeout has been exceeded for the message, so it has been abandoned ({})", packet.getMessage(), transportException);
         }
 
-        IotHubStatusCode errorCode = (transportException instanceof IotHubServiceException) ?
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:1453]: A more exhaustive version of this transportException to status code mapping exists elsewhere, so removing this incomplete mapping

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java [hunk 4]
```diff
@@ -316,6 +316,15 @@ else if (message != null)
                         {
                             clientException = e.toIotHubClientException();
                         }
+                        else
```
Comment (brycewang-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:319]: Do you think it's necessary to add a comment here describing which scenario for instance can be caught by this block?

Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:319]: Good idea, sure

**Review by brycewang-microsoft — APPROVED** [review]

**Pull request merged**